MNE/disp3D/engine/model/3dhelpers/renderable3Dentity.cpp

```diff
@@ -88,56 +88,103 @@ using namespace Qt3DRender;
 
-Renderable3DEntity::Renderable3DEntity(const MatrixX3f& tMatVert, const MatrixX3f& tMatNorm, const MatrixX3i& tMatTris, const MatrixX3f& tMatColors, Qt3DCore::QEntity* parent)
-: Qt3DCore::QEntity(parent)
-, m_pCustomMesh(new CustomMesh(tMatVert, tMatNorm, tMatTris, tMatColors))
-, m_pTransform(new Qt3DCore::QTransform())
-, m_fRotX(0.0f)
-, m_fRotY(0.0f)
-, m_fRotZ(0.0f)
+Renderable3DEntity::~Renderable3DEntity()
 {
-    this->addComponent(m_pCustomMesh);
-    this->addComponent(m_pTransform);
+    //releaseNode(this);
 }
 
 
+////*************************************************************************************************************
+
+//void Renderable3DEntity::releaseNode(Qt3DCore::QNode* node)
+//{
+//    if (QEntity* entity = dynamic_cast<QEntity*>(node)) {
+//        QComponentVector components = entity->components();
+
+//        foreach (QComponent* component, components) {
+//            entity->removeComponent(component);
+//            delete component;
+//        }
+//    }
+
+//    QNodeVector nodes = node->childNodes();
+
+//    foreach (QNode* nodeIt, nodes) {
+//        releaseNode(nodeIt);
+//        delete nodeIt;
+//    }
+//}
+
+
 //*************************************************************************************************************
 
-Renderable3DEntity::~Renderable3DEntity()
+void Renderable3DEntity::setTransform(const Qt3DCore::QTransform& transform)
 {
+    if(m_pTransform) {
+        m_pTransform->setMatrix(transform.matrix());
+    }
 }
 
 
 //*************************************************************************************************************
 
-void Renderable3DEntity::setTransform(QPointer<Qt3DCore::QTransform> pTransform)
+void Renderable3DEntity::applyTransform(const Qt3DCore::QTransform& transform)
 {
     if(m_pTransform) {
-        this->removeComponent(m_pTransform);
-        m_pTransform = pTransform;
-        this->addComponent(m_pTransform);
+        m_pTransform->setMatrix(m_pTransform->matrix() * transform.matrix());
     }
 }
 
 
 //*************************************************************************************************************
 
-QPointer<CustomMesh> Renderable3DEntity::getCustomMesh()
+void Renderable3DEntity::setMaterialParameter(QVariant data, QString sParameterName)
 {
-    return m_pCustomMesh;
+    //Look for all materials and set the corresponding parameters
+    QComponentVector vComponents;
+
+    //Search in this item's components
+    if(QEntity* pEntity = dynamic_cast<QEntity*>(this)) {
+        vComponents = pEntity->components();
+
+        for(int j = 0; j < vComponents.size(); ++j) {
+            if(QMaterial* pMaterial = dynamic_cast<QMaterial*>(vComponents.at(j))) {
+                for(int i = 0; i < pMaterial->effect()->parameters().size(); ++i) {
+                    if(pMaterial->effect()->parameters().at(i)->name() == sParameterName) {
+                        pMaterial->effect()->parameters().at(i)->setValue(data);
```

Comment:
I only use [] if I want to have a non const reference returned. Since, I do not need a const ref here and do not want to some mess with the data I go with .at()